include/nbl/builtin/hlsl/bxdf/base/cook_torrance_base.hlsl

```diff
@@ -246,10 +260,20 @@ struct SCookTorrance
         bool transmitted = math::partitionRandVariable(reflectance, z, rcpChoiceProb);
 
         ray_dir_info_type V = interaction.getV();
-        const vector3_type _N = interaction.getN();
+        const vector3_type H = hlsl::mul(interaction.getFromTangentSpace(), localH);
         Refract<scalar_type> r = Refract<scalar_type>::create(V.getDirection(), H);
         const scalar_type LdotH = hlsl::mix(VdotH, r.getNdotT(rcpEta.value2[0]), transmitted);
-        cache = anisocache_type::createPartial(VdotH, LdotH, hlsl::dot(_N, H), transmitted, rcpEta);
+
+        // fail if samples have invalid paths
+        const scalar_type viewShortenFactor = hlsl::mix(scalar_type(1.0), rcpEta.value[0], transmitted);
+        const scalar_type NdotL = localH.z * (VdotH * viewShortenFactor + LdotH) - NdotV * viewShortenFactor;
```

Comment:
the equation I gave you for this is wrong, see https://github.com/Devsh-Graphics-Programming/Nabla/pull/930#discussion_r2444561279

it probably explains why your BSDF generate is wrong.